### DISCUSSION TITLE
Thoughts about PEP 712

### DISCUSSION THREAD

**Discussion by samuelcolvin** [pydantic/pydantic#8952] [General]
My brief thoughts about [PEP 712](https://peps.python.org/pep-0712/), some other members of the @pydantic/pydantic-maintainers team might have other thoughts.

Overall I don't think PEP 712 would be very useful to Pydantic, it looks like it would only really be useful with vanilla dataclasses. I don't think it's very sensible to state in the introduction it helps Pydantic without askink the authors - it doesn't!

Overall, I would strongly prefer this PEP was rejected or rewritten to be more widely usable.

There are two main problems with it:

## Not `Annotated`

It's seems like a big missed opportunity to define `converter` using `dataclasses.Field` and `dataclass_transform` - surely it should be defined via `Annotated` - e.g. `Annotated[str | None, Converter(str_or_none)]`.

This would still be (mostly) consumable by dataclasses, but would have two significant advantages:
* if could be used by function decorators
* it could be used in our [`TypeAdapter`](https://docs.pydantic.dev/latest/concepts/type_adapter/) if `TypeForm` gets accepted - see [this discussion](https://discuss.python.org/t/is-annotated-compatible-with-type-t/43898/22)
* Using `Annotated` would mean converters could work on recursive types e.g. `list[Annotated[str | None, Converter(str_or_none)]]` which as far as I can tell is not even considered by the PEP, the alternative to this would be defining `list_int_converter`, `list_str_converter`, `set_int_converter`, `set_str_converter` etc. etc. (instead of just defining `int_converter` and `set_converter`)

(Of course the recursive case is more problematic for the dataclasses implementation since it would require them to recurse through types apply validators which is a can of worms, but I don't the burden of implementation on dataclasses is a good reason to accept the PEP while it's still basically a toy)

## No way to set library defaults

**(major problem for pydantic users)**

PEP 712 puts the burden on the end developer to define the converter(s) used for each field - it would be massively more impactful if we could somehow tell type checkers about the default converters Pydantic applies in both lax and strict mode.

We already have these defined in code, they're used to build [this part of our docs](https://docs.pydantic.dev/latest/concepts/conversion_table/) but we define them [as code](https://github.com/pydantic/pydantic/blob/b09e06cedd18fe8f3bd5983682e81ffe72bb63e9/docs/plugins/conversion_table.py#L106) precisely in case there is ever an opportunity like this to tell type checkers about our type conversions.

Instead of manually setting the converter for each field, I would propose something like this:

```py
@dataclass_transform(
    kw_only_default=True,
    type_converts={
        int: pydantic_int_converter,
        float: pydantic_float_converter,
        str: pydantic_str_converter,
        ...
   }
)
class BaseModel:
    ...
```

These types would then be used by the type checker to understand how pydantic converts values, including on recursive types, e.g. `int: pydantic_int_converter` would be applicable to both `age: int` and `ages: list[int]`.

`dataclass_transform(type_convertes)` could be used for our `BaseModel` and `dataclass`, we'd need another way to tell the type checker about these conversions for `TypeAdapter` and [`validate_call`](https://docs.pydantic.dev/latest/concepts/validation_decorator/), but that could come later.

I'm not precious about the mechanism or syntax, the point is that we'd really want to be able to define converters simply by using `BaseModel`, or our variant of `dataclass` - without requiring the user to manually add `dataclasses.field(converter=pydantic_str_converter))` in hundreds of positions, which would be error prone anyway.

**Comment by davidhewitt** [comment]
I think there's some complexity here related to interactions with `validate_assignment` too, if we support converter for pydantic dataclasses (or even `BaseModel`).

As proposed I understand this converter of PEP 712 will also run on assignment. This would mean that there is some level of coupling between `_init_` and assignment. We have `validate_assignment` to control whether the validation runs on assignment. With `converter` then part of the assignment pipeline (namely, the `converter`) has to run on assignment even when `validate_assignment` is disabled.

Probably a surmountable problem that's just an implementation detail for Pydantic to solve. Still, it seems a shame that this `converter` field is a bit all-or-nothing (maybe dataclass could accept a `convert_assignment` option, if this PEP were accepted?)

**Comment by Viicos** [comment]
> It's seems like a big missed opportunity to define `converter` using `dataclasses.Field` and `dataclass_transform` - surely it should be defined via `Annotated` - e.g. `Annotated[str | None, Converter(str_or_none)]`.

As a slightly unrelated side note, `Annotated` is currently not supported at all by [^1]:
- stdlib dataclasses: `a: Annotated[int, field(kw_only=True)]` is _not_ supported, `a` will be valid as a positional argument
- `@dataclass_transform`: taking Pydantic as an example: `a: Annotated[int, Field(alias="b")]` will _not_ make type checkers infer that `b` should be provided instead of `a` (which can be confusing).

It would be nice for `@dataclass_transform` to support `Annotated`, as it is a handy pattern. However, I think its intended usage is primarily to extend the stdlib dataclasses behavior to 3rd party libraries, so `Annotated` would first have to be usage in the stdlib.

[^1]: Examples make use of `kw_only` and `alias`, but this applies to all `field/Field` arguments.
